src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -186,6 +188,16 @@ public void prepareWorkerForGracefulShutdown() {
     }
   }
 
+  private void exitPostPipelineFailure() {
+    // Shutdown the worker if a pipeline fails. By means of the spring lifecycle
+    // hooks - e.g. the `PreDestroy` hook here - it will attempt to gracefully
+    // spin down the pipeline
+    // Consider defining exit codes to better afford out of band instance
+    // recovery
+    int code = SpringApplication.exit(springContext, () -> 1);
```

Comment:
This line needs a timeout incase it gets stuck during the shutdown workflow:
```
        at java.lang.Object.wait0(java.base@20/Native Method)
        - waiting on <no object reference available>
        at java.lang.Object.wait(java.base@20/Object.java:366)
        at java.lang.Thread.join(java.base@20/Thread.java:2109)
        - locked <0x0000000740a01fc0> (a java.lang.Thread)
        at build.buildfarm.worker.Pipeline.join(Pipeline.java:182)
        at build.buildfarm.worker.Pipeline.close(Pipeline.java:96)
        at build.buildfarm.worker.shard.Worker.stop(Worker.java:647)
        at java.lang.invoke.LambdaForm$DMH/0x00000008011f0000.invokeVirtual(java.base@20/LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$MH/0x0000000801516000.invoke(java.base@20/LambdaForm$MH)
        at java.lang.invoke.Invokers$Holder.invokeExact_MT(java.base@20/Invokers$Holder)
        at jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(java.base@20/DirectMethodHandleAccessor.java:154)
        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(java.base@20/DirectMethodHandleAccessor.java:104)
        at java.lang.reflect.Method.invoke(java.base@20/Method.java:578)
        at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:389)
        at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeDestroyMethods(InitDestroyAnnotationBeanPostProcessor.java:347)
        at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeDestruction(InitDestroyAnnotationBeanPostProcessor.java:177)
        at org.springframework.beans.factory.support.DisposableBeanAdapter.destroy(DisposableBeanAdapter.java:197)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroyBean(DefaultSingletonBeanRegistry.java:587)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingleton(DefaultSingletonBeanRegistry.java:559)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.destroySingleton(DefaultListableBeanFactory.java:1163)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingletons(DefaultSingletonBeanRegistry.java:520)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.destroySingletons(DefaultListableBeanFactory.java:1156)
        at org.springframework.context.support.AbstractApplicationContext.destroyBeans(AbstractApplicationContext.java:1106)
        at org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:1075)
        at org.springframework.context.support.AbstractApplicationContext.close(AbstractApplicationContext.java:1021)
        - locked <0x0000000740a09f08> (a java.lang.Object)
        at org.springframework.boot.SpringApplication.close(SpringApplication.java:1365)
        at org.springframework.boot.SpringApplication.exit(SpringApplication.java:1352)
```

Reply:
In the latest revision, I've added 2 mechanisms to guard against the condition which I hit in production on the first version of this PR. The [deadlining of a shutdown ensures](https://github.com/bazelbuild/bazel-buildfarm/pull/1309/files#diff-43afc2d02d6395012fade1d0104600664bf1747d9ed8fbffe48f666f396f1f92R202) provides better guarantees we don't exceed a set threshold in this case.